### PR TITLE
Documentation on search space and parallel/sequential tuning

### PR DESCRIPTION
**Pull request by qingyun-wu**
Documentation update:
- Guidelines on parallel vs sequential tuning
- Guidelines on hyperparameter search space

### CODE REVIEW

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 1]
```diff
@@ -383,6 +383,26 @@ automl.fit(X_train, y_train, n_jobs=4, n_concurrent_trials=4)
...
+
+You can use the following way to roughly estimate the wall-clock time in parallel tuning and sequential tuning: To finish $N$ trials of hyperparameter tuning, i.e., run $N$ hyperparameter configurations, the total wall-clock time needed is $N/k*(SingleTrialTime + Overhead)$, in which $SingleTrialTime$ is the trial time to evaluate a particular hyperparameter configuration, $k$ is the scale of parallelism, e.g., the number of parallel CPU/GPU cores, and $Overhead$ is the computation overhead.
+
+In sequential tuning, $k=1$, and in parallel tuning $k>1$. This may suggest that parallel tuning has a shorter wall-clock time. But it is not always the case considering the other two factors $SingleTrialTime$, and $Overhead$:
+
+- The $Overhead$ in sequential tuning is typically neglectable and in parallel tuning is typically large.
```
Comment (sonichi) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:396]:
```suggestion
- The $Overhead$ in sequential tuning is typically neglectable and in parallel tuning is relatively large.
```

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 2]
```diff
@@ -383,6 +383,26 @@ automl.fit(X_train, y_train, n_jobs=4, n_concurrent_trials=4)
...
+
+- You can also try to reduce the $SingleTrialTime$ to reduce the wall-clock time in sequential tuning: For example, by increasing the resource consumed by a single trial (distributed or multi-thread training), you can reduce $SingleTrialTime$. One concrete example is to use the `n_jobs` parameter that sets the number of threads the fitting process can use in many scikit-learn style algorithms.
+
+**(2) Considerations on randomness.**
+
+Potential reasons that cause randomness:
+1. Parallel tuning: In the case of parallel tuning, the order of trials' finishing time is no longer deterministic. This non-deterministic order, combined with sequential HPO algorithms, leads to a different hyperparameter tuning trajectory.
```
Comment (sonichi) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:403]:
```suggestion
1. Parallel tuning: In the case of parallel tuning, the order of trials' finishing time is no longer deterministic. This non-deterministic order, combined with sequential HPO algorithms, leads to a non-deterministic hyperparameter tuning trajectory.
```

File: website/docs/Use-Cases/Tune-User-Defined-Function.md [hunk 3]
```diff
@@ -74,16 +74,43 @@ from flaml import tune
...
+
+ If it is a categorical hyperparameter, then you should use `tune.choice(possible_choices)` in which `possible_choices` is the list of possible categorical values of the hyperparameter. For example, if you are tuning the optimizer used in model training, and the candidate optimizers are "sgd" and "adam", you should specify the search space in the following way:
+```python
+{
+    "optimizer": tune.choice(["sgd", "adam"]),
+}
+```
+- **Numerical hyperparameter**
+
+If it is a numerical hyperparameter, you need to know does it take integer values or float values. In addition, you need to know:
```
Comment (sonichi) [inline website/docs/Use-Cases/Tune-User-Defined-Function.md:97]:
```suggestion
If it is a numerical hyperparameter, you need to know whether it takes integer values or float values. In addition, you need to know:
```

File: website/docs/Use-Cases/Tune-User-Defined-Function.md [hunk 4]
```diff
@@ -74,16 +74,43 @@ from flaml import tune
...
+- The range of valid values, i.e., what are the lower limit and upper limit of the hyperparameter value?
+- Do you want to sample in linear scale or log scale? It is a common practice to sample in the log scale if the valid value range is large and the evaluation function changes more regularly with respect to the log domain. For example, learning rate. Otherwise, use the linear scale.
```
Comment (sonichi) [inline website/docs/Use-Cases/Tune-User-Defined-Function.md:99]: It's better to give two different search ranges of the learning rate which are suitable for log scale and linear scale respectively. Not all search ranges are suitable for log scale.
Also, it's better to give examples for the other two questions.

Comment (qingyun-wu) [reply website/docs/Use-Cases/Tune-User-Defined-Function.md]: Do you have an example in which the linear scale is more suitable than the log scale for learning rate tuning?

File: website/docs/Use-Cases/Tune-User-Defined-Function.md [hunk 5]
```diff
@@ -94,14 +94,32 @@ The domain specifies a *type* and *valid range* to sample parameters from. Suppo
...
+
 - Do you have quantization granularity requirements?
 
+When you have a desired quantization granularity for the hyperparameter change, you can use `tune.qlograndint` or `tune.qloguniform` to realize the quantization requirement.
+The following code example helps you realize the suggested practice for learning rate tuning from *Goodfellow, I., Bengio, Y., & Courville, A. (2016). Deep learning. MIT Press* (Page 434): "picking values approximately on a logarithmic scale, e.g., a learning rate taken within the set {.1, .01, 10−3, 10−4 , 10−5}"
+```python
+{
+"learning_rate": tune.qloguniform(lower=1e-5, upper=0.1, q=0.1),
+}
+```
```
Comment (sonichi) [inline website/docs/Use-Cases/Tune-User-Defined-Function.md:117]: Both examples are problematic. For the first one, learning_rate doesn't always need log scale. It depends on the search range. You can provide two examples of search ranges, one with log scale and another with linear scale.

For the second example, q=0.1 means round to an integer multiple of 0.1s, e.g., 0.1, 0.2, 0.3, ...

Comment (qingyun-wu) [reply website/docs/Use-Cases/Tune-User-Defined-Function.md]: You are right. I revised the examples.

**Review by sonichi — APPROVED** [review]

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 6]
```diff
@@ -383,6 +383,26 @@ automl.fit(X_train, y_train, n_jobs=4, n_concurrent_trials=4)
 ```
 flaml will perform 4 trials in parallel, each consuming 4 CPU cores. The parallel tuning uses the [BlendSearch](Tune-User-Defined-Function##blendsearch-economical-hyperparameter-optimization-with-blended-search-strategy) algorithm.
 
+#### **Guidelines on parallel vs sequential tuning**
+
+**(1) Considerations on wall-clock time.**
+
+One common motivation for parallel tuning is to save wall-clock time. When sequential tuning and parallel tuning can achieve a similar wall-clock time, sequential tuning should be preferred. This is a rule of thumb when the HPO algorithm is sequential in nature (e.g., the HPO algorithms CFO and BS in FLAML, and most BO algorithms). Sequential tuning allows the HPO algorithms to take advantage of historical trial results. Then the question is **How to estimate the wall-clock-time needed by parallel tuning and sequential tuning**?
```
Comment (liususan091219) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:390]:
```suggestion
One common motivation for parallel tuning is to save wall-clock time. When sequential tuning and parallel tuning achieve a similar wall-clock time, sequential tuning should be preferred. This is a rule of thumb when the HPO algorithm is sequential by nature (e.g., Bayesian Optimization and FLAML's HPO algorithms CFO and BS). Sequential tuning allows the HPO algorithms to take advantage of the historical trial results. Then the question is **How to estimate the wall-clock-time needed by parallel tuning and sequential tuning**?
```

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 7]
```diff
@@ -383,6 +383,26 @@ automl.fit(X_train, y_train, n_jobs=4, n_concurrent_trials=4)
...
+
+You can use the following way to roughly estimate the wall-clock time in parallel tuning and sequential tuning: To finish $N$ trials of hyperparameter tuning, i.e., run $N$ hyperparameter configurations, the total wall-clock time needed is $N/k*(SingleTrialTime + Overhead)$, in which $SingleTrialTime$ is the trial time to evaluate a particular hyperparameter configuration, $k$ is the scale of parallelism, e.g., the number of parallel CPU/GPU cores, and $Overhead$ is the computation overhead.
+
+In sequential tuning, $k=1$, and in parallel tuning $k>1$. This may suggest that parallel tuning has a shorter wall-clock time. But it is not always the case considering the other two factors $SingleTrialTime$, and $Overhead$:
+
+- The $Overhead$ in sequential tuning is typically neglectable and in parallel tuning is relatively large.
```
Comment (liususan091219) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:396]:
```suggestion
- The $Overhead$ in sequential tuning is typically negligible; while in parallel tuning, it is relatively large.
```

**Review by liususan091219 — CHANGES_REQUESTED** [review]

**Review by liususan091219 — APPROVED** [review]
